components/net/fetch/methods.rs

```diff
@@ -264,10 +264,33 @@ pub fn main_fetch(request: Rc<Request>,
         }
 
         // Step 16
-        // TODO this step (CSP/blocking)
+        // TODO Blocking for CSP, mixed content, MIME type
+        let blocked_error_response;
+        let internal_response = if !response.is_network_error() && should_block_nosniff(&request, &response) {
+            // Defer rebinding result
+            blocked_error_response = Response::network_error(NetworkError::Internal("Blocked by nosniff".into()));
+            &blocked_error_response
+        } else {
+            internal_response
+        };
 
         // Step 17
-        if !response.is_network_error() && (is_null_body_status(&internal_response.status) ||
+        // We use `internal_response` since we did not mutate `response` in the previous step.
+        // Steps 14-18 may modify a response to become a network error, but not the reverse.
+        // * Step 14: `internal_response.is_network_error()` <-> `response.is_network_error()`.
+        //  `response.internal_response` may only be a non-network error type of the following:
+        //  - basic filtered response
+        //  - A CORS filtered response
+        //  - An opaque filtered response
+        //  - An opaque-redirect filtered response
+        //  - None
+        // * Step 15: No pertinent change.
+        // * Step 16:
+        //  - If *response* is already a network error, *response* and *internalResponse* are not modified.
+        //  - If *response* becomes a network error, *internalResponse* also become a network error.
+        //  - Otherwise  *response* and *internalResponse* are not modified.
+        // Therefore: `internal_response.is_network_error() == response.is_network_error().`
+        if !internal_response.is_network_error() && (is_null_body_status(&internal_response.status) ||
```

Comment:
We really should avoid writing long comments because it would easily confuse future maintainers. Perhaps the easier way would be to check both `response` and `internal_response` and see if any of the two is a network error, and then leave a short comment there saying that `internal_response` needs to be checked because step 16 did not modify `response` to be a network error.